Updates to uattributes and MQTT5 specifications (#204)
eclipse-uprotocol/up-spec: up-l1/mqtt_5.adoc

For more information, please refer to https://mqtt.org/

This document will discuss the uTransport implementation on MQTT 5 for two well known use cases:
1. *Device-2-Device (D2D) Communication:* This is when devices (through the streamer) was to send messages to other devices through a cloud gateway (broker). 
2. *uE-2-uE Communication:* When MQTT is used as a local software bus for uEs to talk to each other through a local broker.

> The reason fo the different configurations is that cloud MQTT brokers often have limitations on the number of topics that can be subscribed to as well as the topic segment lengths so we need to ensure that message routing is as efficient as possible.

Reviewer: ```suggestion
The reason for the different configurations is that cloud MQTT brokers often have limitations on the number of topics that can be subscribed to as well as the topic segment lengths so we need to ensure that message routing is as efficient as possible.
```

== MQTT5 Header

MQTT 5 supports custom header fields, and we leverage this to map the `UAttributes` values into the `UserProperty` MQTT header in string format.

[.specitem,oft-sid="req~up-transport-mqtt5-attributes~1",oft-needs="impl,utest"]
--
* `UserProperty` Keys *MUST* map to the UAttributes protobuf field numbers.

> --

Reviewer: FMPOV there is no need for this as you are explicitly listing all attributes in the following table, along with the keys to use.

However, what we probably should define is that only non-empty attribute values should be mapped to MQTT header fields.

> * Only non-empty attribute values *MUST* be mapped to MQTT header `UserProperty` keys.

Reviewer: ```suggestion
* All non-empty attribute values *MUST* be mapped to MQTT header *User Properties* using the keys defined below.
* Empty attribute values *MUST NOT* be mapped.
```

--

|===

All uAttributes are mapped to a MQTT header UserProperty, where the key is the UAttributes protobuf field number. The value is a string representation of the UAttributes field. Only UAttributes that are used in the message are included in the MQTT header. If a UAttributes field is not present in the header, than it is considered not used when recompiling the UAttributes.

== Payload Encoding

[.specitem,oft-sid="req~up-transport-mqtt5-payload-encoding~1",oft-needs="impl,utest"]
--
* The MQTT payload **MUST** be the `UMessage.payload` field, which is a byte array to reduce size.
--

== MQTT5 Topics

The MQTT topic a message is published on utilizes the source and sink UUri fields. The topic is dependent on the use case for the transport implementation that will be discussed below.

=== uE-2-uE Communication

When a message is sent from one uE to another uE, the topic is constructed using the source and sink UUri fields. The topic is constructed as follows using the string representation of a `UUri`:

[.specitem,oft-sid="req~up-transport-mqtt5-ue2ue-topic~1",oft-needs="impl,utest"]
--

> `{UAttributes.source}/{UAttributes.sink}/`

Reviewer: what is the purpose of the trailing `/` character?

Reviewer: also, what you have specified means that the topic always starts with a `/` character, which is unusual from my experience

Author: Removed the trailing '/'. Topic won't start with '/' as every message will always have a source but not always a sink.

--

[.specitem,oft-sid="req~up-transport-mqtt5-ue2ue-topic-nosink~1",oft-needs="impl,utest"]
--
If the messages does not have a sink `UUri`, then the sink portion of the MQTT5 topic *MUST* be omitted.
--

==== Examples

.uE-2-uE Communication Topics
[cols="1,2,2,4"]
|===
| Type| source URI | sink URI | MQTT5 Topic

| *Request* | `//device1/AB34/1/0` | `//device1/43BA/1/2` | `device1/AB34/1/0/device1/43BA/1/2`
| *Response* | `//device1/43BA/1/2` | `//device1/AB34/1/0` | `device1/43BA/1/2/device1/AB34/1/0`
| *Publish* | `//device1/AB34/1/8000` | None | `device1/AB34/1/8000`

> | *Notification* | `//device1/43BA/1/8001` | `//device1/AB34/1/0` | `device1/43BA/1/8001/device1/AB34/1/0`

Reviewer: these examples do not match the topic pattern defined above

Reviewer: with a source of `//device1/43BA/1/8001` and a sink of `//device1/AB34/1/0` and a pattern of `$(source}/${sink}`, the topic should be `//device1/43BA/1/8001///device1/AB34/1/0`, shouldn't it? However, in the table, you are using `device1/43BA/1/8001/device1/AB34/1/0`.

Even considering that the string representation of a UUri without an authority name actually is not `//device1/43BA/1/8001` but `/device1/43BA/1/8001`, applying the pattern would yield `/device1/AB34/1/8000//device1/AB34/1/0` ...

Author: OK I see your point. We cannot simply append the strings so I will go back to listing each attribute independently.

Author: done

For more information, please refer to https://mqtt.org/

This document will discuss the uTransport implementation on MQTT 5 for two well known use cases:

> 1. *Device-2-Device (D2D) Communication:* This is when devices (through the streamer) was to send messages to other devices through a cloud gateway (broker). 

Reviewer 2: From my reading then a uStreamer would only need put the MQTT5 transport into a special mode and do nothing else special.

Is that correct?

| *Response* | `//device1/43BA/1/2` | `//device1/AB34/1/0` | `device1/43BA/1/2/device1/AB34/1/0`
| *Publish* | `//device1/AB34/1/8000` | None | `device1/AB34/1/8000`
| *Notification* | `//device1/43BA/1/8001` | `//device1/AB34/1/0` | `device1/43BA/1/8001/device1/AB34/1/0`
|===

.uE-2-uE Communication UTransport::registerListener() Examples
[cols="1,1,1,1"]
|===
| Use Case | source filter | sink filter | MQTT Subscription

| Single Publish Topic | `//device1/AB34/1/8000` | None | `device1/AB34/1/8000`
| Incoming requests for a Method | empty | `//device1/AB34/1/12CD` | `\+/+/\+/+/device1/AB34/1/12CD`
| Any Notifications or RPC Responses | empty | //device1/AB34/1/0 | `\+/+/\+/+/device1/AB34/1/0`

|===

=== D2D Communication

When MQTT5 (broker) is used for D2D communication, the topic shall consist of only the authority portion of the source and sink `UUri` from `UAttributes` as shown below:

[.specitem,oft-sid="req~up-transport-mqtt5-d2d-topic~1",oft-needs="impl,utest"]
--
`{source.authority_name}/{sink.authority_name}`
--

==== Examples

===== Registering Listener to Receive All Messages 

> `UTransport::registerListener(ANY, getSource())` where getSource(0 returns the local device's UUri, this translates into the MQTT subscribe to topic `+/{my_source_authority_name}`

Reviewer 2: ```suggestion
`UTransport::registerListener(ANY, getSource())` where getSource() returns the local device's UUri, this translates into the MQTT subscribe to topic `+/{my_source_authority_name}`
```
